### PR TITLE
Hierarchy

### PR DESCRIPTION
**Pull request by joeflack4**
Updates
```
    Backend
    - Finished off hierarchy portion of route cr-hierarchy
    - Added functions for building hierarchy to utils
    - Misc: Codestyle updates
```

### CODE REVIEW

File: backend/app.py [hunk 1]
```diff
@@ -164,83 +166,28 @@ def cset_members_items(codeset_ids: List[int] = None, con=CON) -> List[LegacyRow
...
-        WHERE hier.depth < 2
-    )
-    SELECT DISTINCT path, depth
-    FROM hier
-    ORDER BY path;
-    """
-    return hier
+    # sql speed: 36-48sec concept_relationship (n=16,971,521). 1.8sec concept_relationship_subsumes_only (n=875,090)
+    t0 = datetime.now()
+    rows: List[LegacyRow] = sql_query(con, "SELECT concept_id_1, concept_id_2 FROM concept_relationship_subsumes_only;")
```
Comment (joeflack4) [inline backend/app.py:178]: FYI regarding these `SELECT` queries:
> sql speed: 36-48sec concept_relationship (n=16,971,521). 1.8sec concept_relationship_subsumes_only (n=875,090)

File: backend/utils.py [hunk 2]
```diff
@@ -1,25 +1,101 @@
...
+        self.__dict__.update(adict)
 
-  def to_dict(self):
-    return self.__dict__
+    def to_dict(self):
+        """Convert to dictionary"""
+        return self.__dict__
+
+
+def get_nested_from_dict(d: Dict, key_path: List):
```
Comment (joeflack4) [inline backend/utils.py:40]: FYI: `get_nested_from_dict()` and `set_nested_in_dict()`
Helper functions for dealing with nested dictionaries. Used by `hierarchify_list_of_parent_kids()`.

File: backend/utils.py [hunk 3]
```diff
@@ -1,25 +1,101 @@
...
+    return reduce(operator.getitem, key_path, d)
+
+
+def set_nested_in_dict(d: Dict, key_path: List, value: Any):
+    """Set nested value in dictionary"""
+    # noinspection PyUnresolvedReferences
+    get_nested_from_dict(d, key_path[:-1])[key_path[-1]] = value
+
+
+def hierarchify_list_of_parent_kids(
```
Comment (joeflack4) [inline backend/utils.py:51]: @Sigfried : `hierarchify_list_of_parent_kids()`: This is the function that does the recursion. I tried a bunch of different ways, but wasn't having trouble with duplicate subtrees, recursion depth, as well as trying to debug. I found this to be the easiest way, and it seems decently performant.

File: test/test_backend.py [hunk 4]
```diff
@@ -19,10 +21,48 @@ def setup(self):
         """always runs first"""
         pass
 
-    def tearDown(self) -> None:
+    def tearDown(self):
         """always runs last"""
         pass
 
+    def test_hierarchify_list_of_parent_kids(self):
```
Comment (joeflack4) [inline test/test_backend.py:28]: @Sigfried: You might want to check out this unit test. The second test case made it a lot easier for me to tell that it was working correctly.

File: backend/utils.py (hunk 3 above)
Comment (Sigfried) [reply backend/utils.py:51]: @joeflack4: it's not working for me. 

http://127.0.0.1:8000/hierarchy?codeset_id=411456218|40061425|484619125|419757429 :
`Time to get concept_relationship_subsumes_only: 0:00:01.422682
Time to traverse: 0:00:00.765243
INFO:     127.0.0.1:53503 - "PUT /datasets/csets HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
[stack trace is basically all FastAPI stuff]
`

What it's trying to return is:
`d
{(411456218, 138056): {}, (411456218, 252341): {}, (411456218, 252658): {}, (411456218, 252942): {}, (411456218, 254980): {}, (411456218, 255841): {}, (411456218, 256448): {}, (411456218, 256716): {}, (411456218, 257581): {}, (411456218, 257583): {}, (411456218, 257905): {}, (411456218, 258564): {}, (411456218, 258780): {}, (411456218, 259043): {}, (411456218, 259055): {}, (411456218, 261048): {}, (411456218, 261325): {}, (411456218, 261889): {}, .... (484619125, 46274059): {}, (484619125, 46274124): {}}`

Every item appears to be a 2-item tuple key and empty value; no apparent hierarchical structure.

File: test/test_backend.py (hunk 4 above)
Comment (Sigfried) [reply test/test_backend.py:28]: @joeflack4: Also not working for me. I can't figure out how to run the tests in the debugger, but from command line I get:
[junk.txt](https://github.com/jhu-bids/TermHub/files/10268087/junk.txt)

Doesn't seem like that test is running at all

Comment (Sigfried) [reply test/test_backend.py:28]: Ok...I couldn't run the tests in the usual way, but just ran them manually in the debugger (line by line) and see that they work. Not sure why I get something different running the endpoint

File: backend/utils.py (hunk 3 above)
Comment (Sigfried) [reply backend/utils.py:51]: Ok, I found the problem. `hierarchy()` had

        selected_concept_ids = get_concept_set_member_ids(codeset_ids)

which returns tuples of (codeset_id, concept_id), but it needed a flat list of concept_ids, so had to specify the single column:

        selected_concept_ids = get_concept_set_member_ids(codeset_ids, column='concept_id')

Comment (Sigfried) [reply backend/utils.py:51]: @joeflack4, not sure if you'll see the above if I resolve the conversation, but I will anyway

**Comment by Sigfried** [comment]
@joeflack4, it's working now (after 4d8a80b), but seeing some differences from old json output:

subtree from old json 

    "255841": {
      "256448": {
        "40481763": null
      },
      "258780": {
        "46270376": null
      },
      "261889": null,
      "765431": null,
      "4137525": {
        "40481763": null,
        "46270376": null
      },
      "4172303": null,
      "4200851": null,
      "4230358": null,
      "4309350": {
        "257905": null,
        "4112826": null
      },
      "37206130": null,
      "42536541": null
    },

same subtree from new code, http://127.0.0.1:8000/hierarchy?codeset_id=411456218|40061425|484619125|419757429

    "255841": {
      "256448": {
        "40481763": {}
      },
      "258780": {}, -- not empty above
      "261889": {},
      "765431": {},
      "4137525": {
        "3187037": {},  -- doesn't appear in list above at all
        "40481763": {},
        "46270376": {}
      },
      "4172303": {},
      "4200851": {},
      "4230358": {},
      "4236366": {},    -- doesn't appear in list above at all
      "4309350": {
        "257905": {},
        "4112826": {}
      },
      "37206130": {},
      "42536541": {}
    }, 

I'm not sure which version is correct, and I haven't compared any further than this. 

Also, I'm not sure how the front end is going to feel about your replacing null (None) with {}. If it doesn't like it, have to decide whether to fix it there or here.

**Comment by Sigfried** [comment]
@joeflack4:

Definitely an error:

    "258780": {}, -- not empty above

Not errors:

    "3187037": {},  -- doesn't appear in list above at all
    "4236366": {},    -- doesn't appear in list above at all

because they're not in any of those concept sets. I'm not sure why the old code captured them. But we eventually like to add options to include additional related codes outside the selected csets.

**Comment by joeflack4** [comment]
Yeah, I didn't have time to look further evaluate why the old and the new are different, and which of the two are problematic and why. But we should check / add a better test for that. 

I sent some messages on Teams about this, including the None vs {}. I can convert to None on the backend or the frontend as well.

**Pull request merged**